test/integration/scheduler/podsready/scheduler_test.go

```diff
@@ -152,6 +170,151 @@ var _ = ginkgo.Describe("SchedulerWithWaitForPodsReady", func() {
 			util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodClusterQ.Name, prodWl)
 			util.ExpectWorkloadsToBeWaiting(ctx, k8sClient, devWl)
 		})
+
+		ginkgo.It("Should allow for a workload to continue if the timeout is not exceeded", func() {
+			ginkgo.By("create the 'prod1' workload")
+			prodWl1 := testing.MakeWorkload("prod1", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodWl1)).Should(gomega.Succeed())
+
+			ginkgo.By("create the 'prod2' workload after a second")
+			time.Sleep(time.Second)
+			prodWl2 := testing.MakeWorkload("prod2", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodWl2)).Should(gomega.Succeed())
+
+			ginkgo.By("checking the 'prod1' workload is admitted and the 'prod2' workload is waiting")
+			util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodClusterQ.Name, prodWl1)
+			util.ExpectWorkloadsToBeWaiting(ctx, k8sClient, prodWl2)
+
+			ginkgo.By("wait a little bit")
+			time.Sleep(200 * time.Millisecond)
+
+			ginkgo.By("verify workload 'prod1' continues to be admitted")
+			util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodClusterQ.Name, prodWl1)
+			util.ExpectWorkloadsToBeWaiting(ctx, k8sClient, prodWl2)
+		})
+
+	})
+
+	var _ = ginkgo.Context("Short PodsReady timeout", func() {
+
+		ginkgo.BeforeEach(func() {
+			beforeEachWithTimeout(3 * time.Second)
+		})
+
+		ginkgo.AfterEach(func() {
+			afterEach()
+		})
+
+		ginkgo.It("Should requeue a workload which exceeded the timeout to reach PodsReady=True", func() {
+			ginkgo.By("create the 'prod1' workload")
+			prodWl1 := testing.MakeWorkload("prod1", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodWl1)).Should(gomega.Succeed())
+
+			ginkgo.By("create the 'prod2' workload after a second")
+			time.Sleep(time.Second)
+			prodWl2 := testing.MakeWorkload("prod2", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodWl2)).Should(gomega.Succeed())
+
+			ginkgo.By("checking the 'prod1' workload is admitted and the 'prod2' workload is waiting")
+			util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodClusterQ.Name, prodWl1)
+			util.ExpectWorkloadsToBeWaiting(ctx, k8sClient, prodWl2)
+
+			ginkgo.By("exceed the timeout for the 'prod1' workload")
+			time.Sleep(podsReadyTimeout)
+
+			ginkgo.By("wait for the 'prod1' workload to be unadmitted")
+			gomega.Eventually(func() *kueue.Admission {
+				gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(prodWl1), prodWl1)).Should(gomega.Succeed())
+				return prodWl1.Spec.Admission
+			}, util.Timeout, util.Interval).Should(gomega.BeNil())
+
+			ginkgo.By("verify the 'prod2' workload gets admitted and the 'prod1' is waiting")
```

Comment:
Could this be flaky?
Because prod1 would be added back to the queue and it has an old StartTime. Maybe there should be a grace period in which prod1 doesn't enter the queue. Or we could delete the Workload object so that the Job controller recreates it with a newer StartTime. This would be good because we know this workload couldn't schedule, so we should minimize the chances of it clogging the queue again.

However, this behavior might make it harder to reset the node selector. WDYT?

Reply:
The test isn't flaky because it first asserts that the other workload is `waiting` (`util.ExpectWorkloadsToBeWaiting(ctx, k8sClient, prodWl2)`. The waiting workload is admitted first as it completes the previous scheduler cycle.

As for the clogging of the queue, deleting a workload sounds unnecessary and does not seem to reflect the intention, IMO. 

I guess another relatively simple solution would be to adjust the queues logic to use max of `CreationTimestamp` and `LastTransitionTime` for Admitted=False condition. Then, the re-admitted workloads would be last again. Let me know what you think, we could also defer it to a follow up PR.

Reply:
That could work, and I'm fine having it in a follow up PR.

Maybe it could be an option about what to do after preemption. WDYT @ahg-g ?

Reply:
Ah right, there is a wait inside the scheduling cycle. Good.

Reply:
For this test, the two workloads have different priorities, so the timestamps don't have an effect on ordering in the queue because we first order by priority.

> Maybe it could be an option about what to do after preemption. WDYT @ahg-g ?

I am not sure. I think preemption and exceeding ready timeout should probably be treated differently. The workload is not at "fault" in the former, but it is in the latter, and so it seems unfair that they get punished when they get preempted. 

Another two criteria we can incorporate in the order:
1) Job size for the preemption case, smaller preempted jobs gets preference over larger ones.
2) Backoffs for the not ready case